### PR TITLE
Fix #6435 - WaterUse:Storage Design Inflow Rate not respected

### LINKED ISSUE #6435: WaterUse:Storage Design Inflow Rate not respected
Issue overview
--------------
The WaterUse:Storage object seems to assume that the requested refill rate is supplied - even where it is limited. 

### Details
Some additional details for this issue (if relevant):
 - Windows 10 
 - EnergyPlus 8.8
 - Unmethours link https://unmethours.com/question/29064/how-to-model-pump-energy-to-refill-waterusestorage-tank/?answer=29085#post-id-29085

Defect files added 
refill rate bug
[in.zip](https://github.com/NREL/EnergyPlus/files/1629037/in.zip)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fix #6435
 - WaterUse:Storage Design Inflow Rate not respected

The problem is actually slightly different than originally stated, it's a problem of float valve control. When it goes below the `ValveOnCapacity`, it fills up. But at the next timestep, regardless of whether it's went above the `ValveOffCapacity` or not, it checks again if it's below `ValveOnCapacity` and determines that it doesn't need to keep filling up.

Defect file on EnergyPlusDevSupport: https://github.com/NREL/EnergyPlusDevSupport/blob/d37b70ea2df0e180cc527b222f827913c81ecf5d/DefectFiles/6000s/6435/in.idf#L9647-L9681

Running the defect file before and after. As you can see, in the before case, the last graph "Storage Tank Volume [m3]", you have the tank filling up only during one timestep, while the `ValveOffCapacity` which is 3m3 isn't achieved. The Timestep in the file is 6, so it fills up between 14:40 and 14:50. In the after case, the tank fills up during one more timestep until it's `ValveOffCapacity` is achieved.

image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataWater.hh [hunk 1]
```diff
@@ -121,6 +121,7 @@ namespace DataWater {
         int OverflowTankSupplyARRID;
         Real64 ValveOnCapacity;  // tank capacity at lower control range [m3]
         Real64 ValveOffCapacity; // tank capacity at upper control range [m3]
+        bool LastTimeStepFilling; // Indicates that tank was filling up at last timestep, to determine whether to continue until ValveOffCapacity
```
Comment (jmarrec) [inline src/EnergyPlus/DataWater.hh:124]: New attribute

File: src/EnergyPlus/DataWater.hh [hunk 2]
```diff
@@ -172,8 +173,10 @@ namespace DataWater {
 
         // Default Constructor
         StorageTankDataStruct()
-            : MaxCapacity(0.0), OverflowMode(Overflow::Unassigned), OverflowTankID(0), OverflowTankSupplyARRID(0), ValveOnCapacity(0.0), ValveOffCapacity(0.0), ControlSupply(ControlSupplyType::Unassigned), GroundWellID(0), SupplyTankID(0), SupplyTankDemandARRID(0), BackupMainsCapacity(0.0), InitialVolume(0.0),
-              MaxInFlowRate(0.0), MaxOutFlowRate(0.0), ThermalMode(TankThermalMode::Unassigned), InitialTankTemp(20.0), TempSchedID(0), AmbientTempIndicator(AmbientTempType::Unassigned),
+            : MaxCapacity(0.0), OverflowMode(Overflow::Unassigned), OverflowTankID(0), OverflowTankSupplyARRID(0), ValveOnCapacity(0.0),
+              ValveOffCapacity(0.0), LastTimeStepFilling(false), ControlSupply(ControlSupplyType::Unassigned), GroundWellID(0), SupplyTankID(0),
```
Comment (jmarrec) [inline src/EnergyPlus/DataWater.hh:177]: Init to false in Ctor

File: src/EnergyPlus/WaterManager.cc [hunk 3]
```diff
@@ -1049,9 +1049,14 @@ namespace WaterManager {
...
         // is tank lower than float valve on capacity and requesting fill from controlled supplier?
         FillVolRequest = 0.0;
-        if ((VolumePredict) < state.dataWaterData->WaterStorage(TankNum).ValveOnCapacity) { // turn on supply to fill tank
+
+        if ( ((VolumePredict) < state.dataWaterData->WaterStorage(TankNum).ValveOnCapacity) ||
+              state.dataWaterData->WaterStorage(TankNum).LastTimeStepFilling )
+        { // turn on supply to fill tank
             FillVolRequest = state.dataWaterData->WaterStorage(TankNum).ValveOffCapacity - VolumePredict;
 
+            state.dataWaterData->WaterStorage(TankNum).LastTimeStepFilling = true;
```
Comment (jmarrec) [inline src/EnergyPlus/WaterManager.cc:1058]: If you go below `ValveOnCapacity`, you start filling, and you set the `LastTimeStepFilling` to true.
**Or** if `LastTimeStepFilling` is already `true`, you keep filling.

File: src/EnergyPlus/WaterManager.cc [hunk 4]
```diff
@@ -1080,6 +1085,11 @@ namespace WaterManager {
         }
 
         state.dataWaterData->WaterStorage(TankNum).ThisTimeStepVolume = state.dataWaterData->WaterStorage(TankNum).LastTimeStepVolume + NetVolAdd;
+        if (state.dataWaterData->WaterStorage(TankNum).ThisTimeStepVolume >= state.dataWaterData->WaterStorage(TankNum).ValveOffCapacity) {
+            state.dataWaterData->WaterStorage(TankNum).LastTimeStepFilling = false;
+        }
+
```
Comment (jmarrec) [inline src/EnergyPlus/WaterManager.cc:1091]: At the end of the iteration, you check whether you are above or equal to `ValveOffCapacity`, in which case you switch `LastTimeStepFilling` to false.

**Review by jmarrec — COMMENTED** [review]
Left a Walkthrough. The unit test is commented and should be self explanatory so I didn't add walkthrough comments.

**Comment by jmarrec** [comment]
`regression.5ZoneWaterSystems` has a difference in `WELL 1:Water System Groundwater Well Requested Volume Flow Rate [m3/s]` which is expected. The failed one is the SolarShadingGPU as usual.

**Comment by mitchute** [comment]
I've confirmed the fix here. Everything looks good. Merging.

**Pull request merged**